repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py

```diff
@@ -345,12 +345,73 @@ def _get_files_owned_by_rpms(context, dirpath, pkgs=None, recursive=False):
     return files_owned_by_rpms
 
 
+def _copy_decouple(srcdir, dstdir):
+    """
+    Copy `srcdir` to `dstdir` while decoupling symlinks.
+
+    What we mean by decoupling the `srcdir` is that any symlinks pointing
+    outside the directory will be copied as regular files. This means that the
+    directory will become independent from its surroundings with respect to
+    symlinks. Any symlink (or symlinks chains) within the directory will be
+    preserved.
+
+    """
+
+    for root, _, files in os.walk(srcdir):
+        for filename in files:
+            relpath = os.path.relpath(root, srcdir)
+            source_filepath = os.path.join(root, filename)
+            target_filepath = os.path.join(dstdir, relpath, filename)
+
+            # Skip and report broken symlinks
+            if not os.path.exists(source_filepath):
+                api.current_logger().warning('File {} is a broken symlink!'.format(source_filepath))
+                continue
+
+            # Copy symlinks to the target userspace
+            pointee = None
+            if os.path.islink(source_filepath):
+                pointee = os.readlink(source_filepath)
+
+            # If source file is a symlink within `src` then preserve it,
+            # otherwise resolve and copy it as a file it points to
+            if pointee is not None and not pointee.startswith(srcdir):
+                # Follow the path until we hit a file or get back to /etc/pki
+                while not pointee.startswith(srcdir) and os.path.islink(pointee):
+                    pointee = os.readlink(pointee)
+
+                # Pointee points to a _regular file_ outside /etc/pki so we
+                # copy it instead
+                if not pointee.startswith(srcdir) and not os.path.islink(pointee):
+                    source_filepath = pointee
+                else:
+                    # pointee points back to /etc/pki
+                    pass
+
+            # Ensure parent directory exists
+            parent_dir = os.path.dirname(target_filepath)
+            run(['mkdir', '-p', parent_dir])
+
+            if os.path.islink(source_filepath):
+                # Preserve the owner and permissions of the original symlink
+                run(['ln', '-s', pointee, target_filepath])
+                run(['chmod', '--reference={}'.format(source_filepath), target_filepath])
```

Comment:
Just a note for you: on other platforms, symlinks can have their own permissions.  So if we had to be portable to non-Linux systems, you might have to use the real files rather than the symlinks here.

Reply:
Thanks! However, using the real files was the original solution which we had to adjust because changes to  certificates were not propagated it the container (as what were previously symlinks are now regular files which are not updated). So I generally need to preserve the symlink here. Not sure if this is what you meant though. Just out of curiosity, how would you solve this on non-Linux systems?

Reply:
<nod> I was a little unclear, I think.  On other platforms (some of the BSDs for sure) symlinks themselves have file permissions.  So if we needed to be portable to those, we'd need to make sure that the chmod was getting its permissions from the actual source file rather than a symlink and we might have to set those file permissions on all of the possible target_filepaths (the actual target file and the symlink chain leading to it.)  Symlinks with permissions is not a POSIX standard, though, so we'd probably need to test on each platform we cared about to see exactly how their chmod treated the permissions.